### PR TITLE
Fetch Order.fulfillmentTime

### PR DESCRIPTION
**Pull request by Velenir**
- Fetches `block.timestamp` for `Trade event` logs to add as `Order.fulfillmentTime`
- Only fetches for minimum blocks

### CODE REVIEW

File: src/custom/state/orders/updater.ts [hunk 1]
```diff
@@ -94,6 +94,22 @@ const constructGetLogsRetry = (provider: Web3Provider) => {
   return getLogsRetry
 }
 
+const buildBlock2DateMap = async (library: Web3Provider, logs: Log[]): Promise<Record<string, Date>> => {
+  // only check unique blocks
+  // hashes are more stable than numbers in case of reorg
```
Comment (anxolin) [inline src/custom/state/orders/updater.ts:99]: good point

**Comment by gnosis-info** [comment]
Travis automatic deployment:
   * **🎩 [Dapp](https://pr66--gpv2swap.review.gnosisdev.com)**: Testing web app

File: src/custom/state/orders/updater.ts [hunk 2]
```diff
@@ -94,6 +94,22 @@ const constructGetLogsRetry = (provider: Web3Provider) => {
   return getLogsRetry
 }
 
+const buildBlock2DateMap = async (library: Web3Provider, logs: Log[]): Promise<Record<string, Date>> => {
```
Comment (anxolin) [inline src/custom/state/orders/updater.ts:97]: nit: maybe move to a util, to make the updater smaller?

File: src/custom/state/orders/updater.ts [hunk 3]
```diff
@@ -149,6 +165,15 @@ export function EventUpdater(): null {
         topics: eventTopics
       })
 
+      // only update lastCheckedBlock if no logs
+      if (logs.length === 0) {
+        // SET lastCheckedBlock = lastBlockNumber
+        dispatch(updateLastCheckedBlock({ chainId, lastCheckedBlock: lastBlockNumber }))
```
Comment (anxolin) [inline src/custom/state/orders/updater.ts:171]: u see :) you wouldn't need to replicate the dispatch and return in the middle, if u modeled the action as I suggested in the other PR.  There would be just a small map for the orders data. U wanted to use the fancy to use the `batchy-thingy`

**Review by anxolin — APPROVED** [review]

File: src/custom/state/orders/updater.ts (hunk 3 above)
Comment (Velenir) [reply src/custom/state/orders/updater.ts:156]: I'll create a few bulk actions tomorrow for orders, blocks

Comment (anxolin) [reply src/custom/state/orders/updater.ts:156]: still is not the end of the world, but this came to mind when reading. I think is nitypicky

File: src/custom/utils/blocks.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { Web3Provider } from '@ethersproject/providers'
+import { Log } from '@ethersproject/abstract-provider'
+
+export const buildBlock2DateMap = async (library: Web3Provider, logs: Log[]): Promise<Record<string, Date>> => {
+  // only check unique blocks
+  // hashes are more stable than numbers in case of reorg
+  const blockHashes = Array.from(new Set(logs.map(log => log.blockHash)))
```
Comment (W3stside) [inline src/custom/utils/blocks.ts:7]: this could be expanded out a bit

Comment (W3stside) [reply src/custom/utils/blocks.ts:7]: also maybe i'm missing it but why make a `Set` from the array and then make it an Array again?

Comment (W3stside) [reply src/custom/utils/blocks.ts:7]: i see,

i just meant sth like
```ts
const dedupedLogs = new Set(logs.map(log => log.blockHash))
const blockHashes = Array.from(dedupedLogs)
```

**Review by W3stside — APPROVED** [review]

File: src/custom/utils/blocks.ts (hunk 4 above)
Comment (Velenir) [reply src/custom/utils/blocks.ts:7]: Oh, that, sure, will do

**Pull request merged**